### ISSUE TITLE
Ordering bug

### ISSUE THREAD

**Issue by maxcountryman** [coriolinus/accept-header#4]
There appears to be a bug in the implementation. For example, consider this test case:

```diff
diff --git a/src/lib.rs b/src/lib.rs
index 22cfc31..074c2be 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -243,4 +243,22 @@ mod tests {
         let expect = &[("application/json", 1.0, None)];
         perform_test(input, expect);
     }
+
+    #[test]
+    fn ordering_bug() {
+        let input = "Accept: image/webp,image/avif,image/jxl,image/heic,image/heic-sequence,video/*;q=0.8,image/png,image/svg+xml,image/*;q=0.8,*/*;q=0.5";
+        let expect = &[
+            ("image/webp", 1.0, None),
+            ("image/avif", 1.0, None),
+            ("image/jxl", 1.0, None),
+            ("image/heic", 1.0, None),
+            ("image/heic-sequence", 1.0, None),
+            ("image/png", 1.0, None),
+            ("image/svg+xml", 1.0, None),
+            ("video/*", 0.8, None),
+            ("image/*", 0.8, None),
+            ("*/*", 0.5, None),
+        ];
+        perform_test(input, expect);
+    }
 }
```

Where implied quality should be `1.0` it's rendering as `None` and for some reason `avif` is ordered before `webp`. This seems like an incorrect parsing of the example `Accept` value (taken from Safari).

**Comment by coriolinus** [comment]
Implied quality is rendered as `None` because it is implied, not explicit. Only explicit qualities are recorded in the struct. That said, the media types iterator in this case does in fact sort all `None` qualities before all lower explicit qualities. That's not a bug, just an implementation detail.

`image/avif` is sorted before `image/webp` because internally a sort is performed, and the final sort parameter is alphabetical by name. This property is important to ensure that semantically equivalent accept headers compare as equal. As no media type would match against two different non-wildcard values, the actual ordering of names at a particular quality-specificity tier is irrelevant.

In other words, your test case made the incorrect assumption that the original ordering of media types in the input is reflected in the output ordering. Changing it to sort same-tier options alphabetically causes it to pass. If you have a use case for which it is important to preserve the ordering from the input, please reopen with an explanation. In the meantime, I am closing this as not a bug.

([This section](https://www.rfc-editor.org/rfc/rfc9110#section-12.5.1-11) of the RFC is my basis for asserting that input ordering is irrelevant; in that example, `text/html` and `text/x-c` are equal in priority. If ordering mattered, `text/html` would have a higher priority.)

**Comment by maxcountryman** [comment]
While the RFC does not explicitly say that equal ordering should prefer the ordering provided in practice this is how we would expect things to behave.

This is a real bug in most practical scenarios and it would be nice to see it addressed.

Furthermore implicit quality should be set to 1.0 as per the spec so I would also consider that a mistake in the implementation.

**Comment by coriolinus** [comment]
If you need a quality for every media type, use the [`quality_factor` method](https://github.com/coriolinus/accept-header/blob/3f3a278edbeb4a8548e272a79b7380f66b68f8a3/src/media_type.rs#L19). Choosing not to explicitly set it within the model means that the model can distinguish between `foo/bar` and `foo/bar; q=1.0`.

Adherence to the specification is not a bug, even if it produces behavior which defies intuition. I will not implement your ordering preference on my own. I'm willing in principle to consider a PR which retains the original ordering somehow, but it's a tricky design challenge:

- For all instances of `Accept`, `accept: foo/*, bar/*` must compare equal to `accept: bar/*, foo/*`. 
- `Eq::eq` should be `O(n)` on the number of media types, at worst.
- The default behavior should minimize the footprint of the struct. 

This suggests a design in which a feature flag is added to this library, disabled by default. When enabled, the struct stores a second list with an index map, like 

```rust
pub struct Accept {
    /// This is always stored in descending order of quality.
    media_types: Vec<MediaType>,
    /// This is a list of indices into `media_types` ordered by tier, but not alphabetizing within a tier.
    ///
    /// Iterating based on this list produces output preferring items which came first in the input list within a tier,
    /// instead of items which come first alphabetically.
    #[cfg(feature = "prefer-leftmost-items")]
    original_sort_indices: Vec<usize>,
}
```

Changes which would be necessary:

- equality checks ignore `original_sort_indices`
- ordering checks ignore `original_sort_indices`
- implement a private `MediaType::incomplete_cmp` function which does tier-based sorting
- implement a private `MediaType::complete_cmp` function which does `self.incomplete_cmp(other).then_with(...)` for the name-sorting logic
- defer the logic in `Ord` to `complete_cmp`, retaining an ascending sort by default
- add an appropriate `cmp::Reverse` to the sort function in the constructor to avoid the `.reverse()` call
- when the feature flag is enabled, add a new function `Accept::media_types_leftmost` which produces a type ordering preferring items on the left of the input within a given tier.

**Comment by maxcountryman** [comment]
It's not "intuition": the browser expects precedence ordering according to how it's been provided when weighting is equal. Your implementation sorts alphabetically when it should simply not sort at all when weighting is equal; the spec does not ask you to sort yet you do--this is a bizarre choice and not to spec.

If you aren't interested in addressing these bugs that's okay, please just say so.

I am no longer using this crate so I'm not able to help directly, but others should be aware of these issues.
